haskell/test/SMT.hs

```diff
@@ -0,0 +1,176 @@
+import Morley.Nettest
+import qualified Morley.Micheline as MM
+import Morley.Nettest.Pure
+
+import Ligo.BaseDAO.Contract
+import Ligo.BaseDAO.Types
+import Test.Ligo.BaseDAO.Common.StorageHelper as StorageHelper
+import Test.Ligo.BaseDAO.Common
+import SMT.BaseDAO.Management
+import SMT.BaseDAO.Proposal
+import SMT.BaseDAO.Token
+import SMT.Gen
+import SMT.Types
+
+-- originate onece and call multiple time
+test_SMT :: TestTree
+test_SMT = testGroup "SMT"
+  [ testProperty "SMT" $
+      withTests 20 $ property $ do
+        PropertyTestInput (inputs, cs) <- forAll genPropertyTestInput
+        nettestTestProp $
+          (uncapsNettestEmulated $ do
+              -- Set initial level
+              advanceLevel (cs & csBlockchainState & bsLevel)
+
+              guardianContract <- originateSimple "guardian" () dummyGuardianContract
+              tokenContract <- originateSimple "TokenContract" [] dummyFA2Contract
+
+              let newCs = addAddresses cs (unTAddress guardianContract) (unTAddress tokenContract)
+
+              -- Originate Dao for nettest
+              let
+                originateData = UntypedOriginateData
+                  { uodName = "BaseDAO"
+                  , uodBalance = toMutez 0
+                  , uodStorage = untypeValue $ toVal $ (newCs & csStorage)
+                  , uodContract = convertContract baseDAOContractLigo
+                  }
+
+              daoUntyped <- originateUntyped originateData
+              let dao = TAddress @Parameter daoUntyped
+
+              callEntrypointAndCompare dao inputs newCs
+
+          )
+          emulatedImplIntegrational
+  ]
+  where
+    -- Reason: guardian address and governance token address are not known until we originate the contract in nettest
+    addAddresses cs guardAddr govAddr = cs
+        { csStorage = (cs & csStorage)
+            { fsStorage = (cs & csStorage & fsStorage)
+                { sGuardian = guardAddr
+                , sGovernanceToken = (cs & csStorage & fsStorage & sGovernanceToken)
+                    { gtAddress = govAddr
+                    }
+                }
+            }
+        }
+
+callEntrypointAndCompare :: TAddress Parameter -> [ContractCall Parameter] -> ContractState -> EmulatedT PureM ()
+callEntrypointAndCompare _ [] _ = pure ()
+callEntrypointAndCompare dao (cc:ccs) cs = do
+  let haskellResult = callBaseDaoHaskell cc cs
+  ligoResult <- callBaseDaoLigo dao cc cs
+
+  -- Note: `fsConfig` can't be compared because using `getStorage` resulted in different
+  -- different lambda.
+  -- Ex: cRejectedProposalSlashValue:
+  -- - from haskell: `Seq (Seq (DROPN (SS (SS SZ))) (PUSH (VNat 0))) Nop`
+  -- - from ligo: ` Seq (AnnPAIR : % %) (Seq (AnnUNPAIR @ @ % %) (Seq (DROPN (SS (SS SZ))) (PUSH (VNat 0))))`
+
+  (   haskellResult & csStorage & fsStorage
+    , haskellResult & csError
+    , cc
+    ) @==
+    ( ligoResult & csStorage & fsStorage
+    -- , ligoResult & csError
+    , []
+    , cc
+    )
```

Comment:
The `[]` is a leftover for my testing, it should be `ligoResult & csError`.

`cc` is compared to itself, because when the comparison is different, it should also printout the ContractCall. Maybe there is a more elegant way to do this.